### PR TITLE
#164541932: Update and delete an article

### PR DESCRIPTION
**Pull request by Dubby20**
#### What does this PR do?
- It creates a user update and delete article

#### Description of Task to be completed?
- Creates a delete and update api endpoint so that users can delete and update their articles.

#### How should this be manually tested?
- Run `npm run start` to start the application.
- Make a `PUT` request to `/api/v1/articles:id` which will enable an authenticated user to update their article
- Make a `DELETE` request to `/api/v1/articles:id` which will enable an authenticated user to delete their article

### What are the relevant pivotal tracker stories?
- [#164541932](https://www.pivotaltracker.com/story/show/164541932)

#### Any background context you want to provide?
- None

#### Screenshots (if appropriate)
- None

#### Questions:
- N/A

### CODE REVIEW

File: controllers/articles.js [hunk 1]
```diff
@@ -34,4 +34,70 @@ export default class ArticleController {
...
+      });
+      return res.status(200).json({
+        success: true,
+        message: 'Article updated successfully',
+        article: result
+      });
+    } catch (error) {
+      return res.status(500).json({
+        success: false,
+        error: [error.message]
```
Comment (chingsley) [inline controllers/articles.js]: Good job @Dubby20. But I think this should be `errors: [...]`. Ie. the plural, not singular.

File: controllers/articles.js [hunk 2]
```diff
@@ -34,4 +34,70 @@ export default class ArticleController {
...
+        }
+      });
+      return res.status(200).json({
+        success: true,
+        message: 'Article deleted successfully'
+      });
+    } catch (error) {
+      return res.status(500).json({
+        success: false,
+        error: [error.message]
```
Comment (chingsley) [inline controllers/articles.js]: `errors:`

File: middleware/validation.js [hunk 3]
```diff
@@ -84,6 +85,38 @@ export const validateLogin = [
...
+  try {
+    const article = await Article.findOne({
+      where: {
+        id
+      }
+    });
+    if (!article) {
+      return res.status(404).json({
+        success: false,
+        error: 'Article not found'
```
Comment (chingsley) [inline middleware/validation.js]: `errors: ['Article not found']`

File: middleware/validation.js [hunk 4]
```diff
@@ -84,6 +85,38 @@ export const validateLogin = [
...
+      return res.status(401).json({
+        success: false,
+        error: 'You are unauthorized to perform this action'
+      });
+    }
+    return next();
+  } catch (error) {
+    return res.status(500).json({
+      success: false,
+      error: 'Article does not exist'
```
Comment (chingsley) [inline middleware/validation.js]: `errors: ['Article does not exist']`

Comment (Dubby20) [reply middleware/validation.js]: Resolved

File: controllers/articles.js (hunk 2 above)
Comment (Dubby20) [reply controllers/articles.js]: Resolved

File: controllers/articles.js (hunk 1 above)
Comment (Dubby20) [reply controllers/articles.js]: Resolved

**Review by chingsley — APPROVED** [review]

File: controllers/articles.js [hunk 5]
```diff
@@ -34,4 +34,70 @@ export default class ArticleController {
...
+    } catch (error) {
+      return res.status(500).json({
+        success: false,
+        errors: [error.message]
+      });
+    }
+  }
+
+  /**
+ * @description - Update an article
```
Comment (olajide1234) [inline controllers/articles.js]: Please review this, should be `Delete an article`

File: routes/v1.js [hunk 6]
```diff
@@ -33,12 +34,32 @@ apiRoutes.route('/verify/:verificationId')
...
     addImages,
     validateArticle,
     returnValidationErrors,
     generateSlug,
-    createArticle);
+    createArticle
+  );
+
+apiRoutes.route('/articles/:id')
+  .put(
```
Comment (olajide1234) [inline routes/v1.js:50]: Please consider using a `patch` method here. Put supposes you are sending the entire resource body. See more here: https://stackoverflow.com/questions/28459418/rest-api-put-vs-patch-with-real-life-examples

File: test/article.spec.js [hunk 7]
```diff
@@ -153,3 +159,121 @@ describe('ARTICLES', () => {
...
+  });
+});
+
+describe('/DELETE articles id', () => {
+  it('should return an error if the user is not the owner of the article', (done) => {
+    chai
+      .request(app)
+      .delete(`/api/v1/articles/${articleId}`)
+      .set('authorization', userToken2)
+      .send(article2)
```
Comment (olajide1234) [inline test/article.spec.js]: Not sure you need to send an article when deleting.

File: test/article.spec.js [hunk 8]
```diff
@@ -153,3 +159,121 @@ describe('ARTICLES', () => {
...
+        done(err);
+      });
+  });
+
+  it('should delete an article', (done) => {
+    chai
+      .request(app)
+      .delete(`/api/v1/articles/${articleId}`)
+      .set('authorization', userToken)
+      .send(article2)
```
Comment (olajide1234) [inline test/article.spec.js]: Not sure you need to send an article when deleting.

File: test/article.spec.js [hunk 9]
```diff
@@ -153,3 +159,121 @@ describe('ARTICLES', () => {
...
+        done(err);
+      });
+  });
+
+  it('should return an error if the article is not found', (done) => {
+    chai
+      .request(app)
+      .delete('/api/v1/articles/eab6fbb6-aeda-4e1b-b4be-3582f51a6d30')
+      .set('authorization', userToken)
+      .send(article2)
```
Comment (olajide1234) [inline test/article.spec.js]: Not sure you need to send an article when deleting.

**Review by olajide1234 — COMMENTED** [review]
Thanks, Dubby! Please take a look at these few comments.

File: test/article.spec.js (hunk 7 above)
Comment (Dubby20) [reply test/article.spec.js]: True. Thanks

File: test/article.spec.js (hunk 8 above)
Comment (Dubby20) [reply test/article.spec.js]: True. Thanks

File: test/article.spec.js (hunk 9 above)
Comment (Dubby20) [reply test/article.spec.js]: True. Thanks

File: controllers/articles.js (hunk 5 above)
Comment (Dubby20) [reply controllers/articles.js]: Alright. Thanks

Comment (Dubby20) [reply controllers/articles.js]: Done.

File: routes/v1.js (hunk 6 above)
Comment (Dubby20) [reply routes/v1.js:50]: Done

File: test/article.spec.js (hunk 7 above)
Comment (Dubby20) [reply test/article.spec.js]: Done

File: test/article.spec.js (hunk 8 above)
Comment (Dubby20) [reply test/article.spec.js]: Done

File: test/article.spec.js (hunk 9 above)
Comment (Dubby20) [reply test/article.spec.js]: Done

**Review by olajide1234 — APPROVED** [review]

File: middleware/validation.js [hunk 10]
```diff
@@ -84,6 +85,38 @@ export const validateLogin = [
...
+        slug
+      }
+    });
+    if (!article) {
+      return res.status(404).json({
+        success: false,
+        errors: ['Article not found']
+      });
+    }
+    const checkUser = article.userId === user.id;
```
Comment (9jaswag) [inline middleware/validation.js]: @Dubby20 there's no need to save this to a variable if you're just using it below.

File: middleware/validation.js [hunk 11]
```diff
@@ -84,6 +85,38 @@ export const validateLogin = [
     .withMessage('Please provide a valid password.'),
 ];
 
+export const validateUser = async (req, res, next) => {
```
Comment (9jaswag) [inline middleware/validation.js]: @Dubby20 this variable name gives the impression that you're just verifying a user, meanwhile, we're verifying that the user is the author of the article. Let's use a more descriptive method name.

**Review by 9jaswag — CHANGES_REQUESTED** [review]

File: controllers/articles.js [hunk 12]
```diff
@@ -34,4 +34,73 @@ export default class ArticleController {
...
+    const {
+      params: { slug }
+    } = req;
+    try {
+      const result = await Article.update({
+        title,
+        description,
+        body,
+        images,
+        userId: id
```
Comment (9jaswag) [inline controllers/articles.js]: @Dubby20 we don't need to update the `userId` since we've verified that the user is the author of the article.

**Review by 9jaswag — CHANGES_REQUESTED** [review]

File: routes/v1.js [hunk 13]
```diff
@@ -33,12 +34,32 @@ apiRoutes.route('/verify/:verificationId')
...
+
+apiRoutes.route('/articles/:slug')
+  .put(
+    Auth.verifyUser, isUserVerified,
+    validateUser,
+    addImages,
+    validateArticle,
+    returnValidationErrors,
+    updateArticle,
+    generateSlug
```
Comment (9jaswag) [inline routes/v1.js:58]: @Dubby20 why do we need to generate a slug here again?

File: routes/v1.js [hunk 14]
```diff
@@ -33,12 +34,32 @@ apiRoutes.route('/verify/:verificationId')
   .get(UserController.verifyAccount);
 
 apiRoutes.route('/articles')
-  .post(Auth.verifyUser, isUserVerified,
+  .post(
+    Auth.verifyUser, isUserVerified,
```
Comment (9jaswag) [inline routes/v1.js]: @Dubby20 let these be on different lines

**Review by 9jaswag — CHANGES_REQUESTED** [review]

File: controllers/articles.js (hunk 12 above)
Comment (Dubby20) [reply controllers/articles.js]: Alright

File: middleware/validation.js (hunk 11 above)
Comment (Dubby20) [reply middleware/validation.js]: Alright. I will change the name to be more descriptive.

File: middleware/validation.js (hunk 10 above)
Comment (Dubby20) [reply middleware/validation.js]: Okay I've removed the variable

File: routes/v1.js (hunk 14 above)
Comment (Dubby20) [reply routes/v1.js]: Alright

File: routes/v1.js (hunk 13 above)
Comment (Dubby20) [reply routes/v1.js:58]: @9jaswag the documentation says we need to generate `slug` when we update the title of an article.
Screenshot 2019-03-18 at 1 59 50 PM

**Review by 9jaswag — APPROVED** [review]
Nice work!

**Comment by 9jaswag** [comment]
@Dubby20 always ensure you update the right labels at every point in time

**Comment by Dubby20** [comment]
Okay. Thanks

**Pull request merged**